src/Commands/CreateRLSPoliciesForTenantTables.php

```diff
@@ -0,0 +1,81 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Stancl\Tenancy\Commands;
+
+use Illuminate\Console\Command;
+use Illuminate\Database\Eloquent\Model;
+use Illuminate\Support\Facades\DB;
+
+/**
+ * Creates and uses RLS policies for tables of models related to a tenant directly (using BelongsToTenant),
+ * or through a parent primary model (using BelongsToPrimaryModel).
+ *
+ * The models are discovered in the directories configured in the Tenancy config ('tenancy.rls.model_directories').
+ *
+ * This command is used with Postgres + single-database tenancy.
+ */
+class CreateRLSPoliciesForTenantTables extends Command
+{
+    protected $signature = 'tenants:create-rls-policies';
+
+    public function handle(): int
+    {
+        DB::transaction(function () {
+            foreach (tenancy()->getTenantModels() as $model) {
+                $this->useRlsOnModel($model);
+            }
+        });
+
+        return Command::SUCCESS;
+    }
+
+    /**
+     * Make model use RLS if it belongs to a tenant directly, or through a parent primary model.
+     */
+    protected function useRlsOnModel(Model $model): void
+    {
+        $table = $model->getTable();
+        $tenantKey = tenancy()->tenantKeyColumn();
+
+        DB::statement("DROP POLICY IF EXISTS {$table}_rls_policy ON {$table}");
+
+        if (tenancy()->modelBelongsToTenant($model)) {
+            DB::statement("CREATE POLICY {$table}_rls_policy ON {$table} USING ({$tenantKey}::TEXT = current_user);");
+
+            $this->enableRls($table);
+
+            $this->components->info("Created RLS policy for table '$table'");
+        }
+
+        if (tenancy()->modelBelongsToTenantIndirectly($model)) {
+            /** @phpstan-ignore-next-line */
+            $parentName = $model->getRelationshipToPrimaryModel();
+            $parentKey = $model->$parentName()->getForeignKeyName();
+            $parentTable = $model->$parentName()->make()->getTable();
+
+            DB::statement("CREATE POLICY {$table}_rls_policy ON {$table} USING (
+                {$parentKey} IN (
+                    SELECT id
+                    FROM {$parentTable}
+                    WHERE ({$tenantKey} = (
+                        SELECT {$tenantKey}
+                        FROM {$parentTable}
+                        WHERE id = {$parentKey}
```

Comment:
Could we use query parameters (`?`) here instead of directly creating a SQL string in PHP and executing it in the database?

~~`$parentKey` *might be* user data I think, so this might potentially open up the application to SQL injection.~~ Actually it's not the *key*, it's the *key name*. So we should probably rename `$tenantKey` and `$parentKey` to `$tenantKeyName` and `$parentKeyName`. But still, using parameters would be much better than what we're doing here if it'd be possible.

There's a lot going on here (compared to eg the DB managers), so even if it'd be secure now, there's a risk some user data might slip in in the future without being noticed.

Ideally we'd change all of these queries to use parameters (including the ones in all DB managers) if possible.

Reply:
Okay, I'll look into that.

Reply:
I renamed te variables, and looked into supplying the query parameters extensively. I haven't found a way to do that with Postgres queries though.

Reply:
I figured out how to use the parameters in the Postgres statements:

https://github.com/archtechx/tenancy/pull/1105/commits/855ebfbd0b835ef322f95305f53174fc026d35d2
https://github.com/archtechx/tenancy/pull/1105/commits/15cdd6b2c1fe042f147cca2beae86fb076f5896d